File: backend/internal/api/handlers/subscriptions.go
package handlers

import (
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"ytdm/backend/internal/api/response"
	"ytdm/backend/internal/apperr"
	"ytdm/backend/internal/jobs"
	"ytdm/backend/internal/music"
	"ytdm/backend/internal/subscriptions"
)

// subscribeRequest is the body of POST /subscriptions.
//
// The artist is named by the provider it was found on. Nothing here converts
// an artist onto a different provider: a Spotify artist stays a Spotify
// subscription and is synced through the Spotify metadata provider.
type subscribeRequest struct {
	Provider         string               `json:"provider"`
	ArtistSourceID   string               `json:"artist_source_id"`
	ArtistName       string               `json:"artist_name"`
	ArtistImageURL   string               `json:"artist_image_url"`
	AutoDownload     *bool                `json:"auto_download"`
	ReleaseFilter    *music.ReleaseFilter `json:"release_filter"`
	DownloadPriority *jobs.Priority       `json:"download_priority"`
}

// updateSubscriptionRequest is the body of PATCH /subscriptions/{id}. An
// omitted field is left untouched.
type updateSubscriptionRequest struct {
	Enabled          *bool                `json:"enabled"`
	AutoDownload     *bool                `json:"auto_download"`
	ReleaseFilter    *music.ReleaseFilter `json:"release_filter"`
	DownloadPriority *jobs.Priority       `json:"download_priority"`
}

// ListSubscriptions answers GET /subscriptions.
func (h *Handlers) ListSubscriptions(w http.ResponseWriter, r *http.Request) {
	if h.deps.Subscriptions == nil {
		response.Fail(w, r, apperr.CodeInternal, "Subscriptions are not available.")
		return
	}
	limit, offset, err := pagination(r, 100, 500)
	if err != nil {
		response.Error(w, r, err)
		return
	}

	list, err := h.deps.Subscriptions.List(r.Context(), subscriptions.ListFilter{
		Provider:       queryString(r, "provider"),
		ArtistSourceID: queryString(r, "artist_source_id"),
		Limit:          limit,
		Offset:         offset,
	})
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.List(w, list, response.Meta{Count: len(list), Limit: limit, Offset: offset})
}

// Subscribe answers POST /subscriptions. Subscribing to an artist that is
// already watched returns the existing subscription rather than a conflict, so
// a repeated request is harmless.
func (h *Handlers) Subscribe(w http.ResponseWriter, r *http.Request) {
	if h.deps.Subscriptions == nil {
		response.Fail(w, r, apperr.CodeInternal, "Subscriptions are not available.")
		return
	}
	var body subscribeRequest
	if err := decodeJSON(r, &body); err != nil {
		response.Error(w, r, err)
		return
	}
	if strings.TrimSpace(body.ArtistSourceID) == "" {
		response.Fail(w, r, apperr.CodeInvalidRequest, "artist_source_id is required.")
		return
	}

	var (
		defaultAutoDownload bool
		defaultPriority     jobs.Priority
		defaultFilter       music.ReleaseFilter
	)
	if h.deps.Settings != nil {
		defaultAutoDownload, defaultPriority, defaultFilter = h.deps.Settings.GetSubscriptionDefaults()
	} else {
		defaultAutoDownload = false
		defaultPriority = jobs.PriorityLow
		defaultFilter = music.DefaultReleaseFilter()
	}

	req := subscriptions.NewSubscription{
		Provider:         body.Provider,
		ArtistSourceID:   body.ArtistSourceID,
		ArtistName:       body.ArtistName,
		ArtistImageURL:   body.ArtistImageURL,
		AutoDownload:     defaultAutoDownload,
		ReleaseFilter:    &defaultFilter,
		DownloadPriority: &defaultPriority,
	}
	if body.AutoDownload != nil {
		req.AutoDownload = *body.AutoDownload
	}
	if body.ReleaseFilter != nil && body.ReleaseFilter.Any() {
		req.ReleaseFilter = body.ReleaseFilter
	}
	if body.DownloadPriority != nil && body.DownloadPriority.Valid() {
		req.DownloadPriority = body.DownloadPriority
	}

	sub, err := h.deps.Subscriptions.Create(r.Context(), req)
	if err != nil {
		response.Error(w, r, err)
		return
	}
	w.Header().Set("Location", "/api/v1/subscriptions/"+sub.ID)
	response.Data(w, http.StatusCreated, sub)
}

// GetSubscription answers GET /subscriptions/{id}, including the report of the
// most recent run when this process still has one.
func (h *Handlers) GetSubscription(w http.ResponseWriter, r *http.Request) {
	if h.deps.Subscriptions == nil {
		response.Fail(w, r, apperr.CodeInternal, "Subscriptions are not available.")
		return
	}
	id := chi.URLParam(r, "id")

	sub, err := h.deps.Subscriptions.Get(r.Context(), id)
	if err != nil {
		response.Error(w, r, err)
		return
	}

	payload := map[string]any{"subscription": sub}
	if result := h.deps.Subscriptions.LastResult(id); result != nil {
		payload["last_result"] = result
	}
	response.Data(w, http.StatusOK, payload)
}

// UpdateSubscription answers PATCH /subscriptions/{id}.
func (h *Handlers) UpdateSubscription(w http.ResponseWriter, r *http.Request) {
	if h.deps.Subscriptions == nil {
		response.Fail(w, r, apperr.CodeInternal, "Subscriptions are not available.")
		return
	}
	var body updateSubscriptionRequest
	if err := decodeJSON(r, &body); err != nil {
		response.Error(w, r, err)
		return
	}

	sub, err := h.deps.Subscriptions.Update(r.Context(), chi.URLParam(r, "id"),
		subscriptions.Update{
			Enabled:          body.Enabled,
			AutoDownload:     body.AutoDownload,
			ReleaseFilter:    body.ReleaseFilter,
			DownloadPriority: body.DownloadPriority,
		})
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.Data(w, http.StatusOK, sub)
}

// DeleteSubscription answers DELETE /subscriptions/{id}. Only the subscription
// goes: everything that was already downloaded stays in the library.
func (h *Handlers) DeleteSubscription(w http.ResponseWriter, r *http.Request) {
	if h.deps.Subscriptions == nil {
		response.Fail(w, r, apperr.CodeInternal, "Subscriptions are not available.")
		return
	}
	if err := h.deps.Subscriptions.Delete(r.Context(), chi.URLParam(r, "id")); err != nil {
		response.Error(w, r, err)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// SyncSubscription answers POST /subscriptions/{id}/sync.
//
// The run happens in the background and the request returns at once. Walking a
// discography takes longer than a request may be held open, and the download
// endpoints already work the same way: the answer is the accepted order, the
// progress arrives on the event stream.
func (h *Handlers) SyncSubscription(w http.ResponseWriter, r *http.Request) {
	if h.deps.Subscriptions == nil {
		response.Fail(w, r, apperr.CodeInternal, "Subscriptions are not available.")
		return
	}
	sub, err := h.deps.Subscriptions.StartSync(r.Context(), chi.URLParam(r, "id"))
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.Data(w, http.StatusAccepted, sub)
}

// ExportSubscriptions answers GET /subscriptions/export with a versioned JSON document.
func (h *Handlers) ExportSubscriptions(w http.ResponseWriter, r *http.Request) {
	if h.deps.Subscriptions == nil {
		response.Fail(w, r, apperr.CodeInternal, "Subscriptions are not available.")
		return
	}

	payload, err := h.deps.Subscriptions.Export(r.Context())
	if err != nil {
		response.Error(w, r, err)
		return
	}

	filename := fmt.Sprintf("ytmdl-subscriptions-%s.json", time.Now().UTC().Format("2006-01-02"))
	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%q", filename))
	response.Data(w, http.StatusOK, payload)
}

// PreviewImportSubscriptions answers POST /subscriptions/import/preview.
func (h *Handlers) PreviewImportSubscriptions(w http.ResponseWriter, r *http.Request) {
	if h.deps.Subscriptions == nil {
		response.Fail(w, r, apperr.CodeInternal, "Subscriptions are not available.")
		return
	}

	var payload subscriptions.ExportPayload
	if err := decodeJSON(r, &payload); err != nil {
		response.Error(w, r, err)
		return
	}

	preview, err := h.deps.Subscriptions.PreviewImport(r.Context(), payload)
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.Data(w, http.StatusOK, preview)
}

// ApplyImportSubscriptions answers POST /subscriptions/import/apply.
func (h *Handlers) ApplyImportSubscriptions(w http.ResponseWriter, r *http.Request) {
	if h.deps.Subscriptions == nil {
		response.Fail(w, r, apperr.CodeInternal, "Subscriptions are not available.")
		return
	}

	var payload subscriptions.ExportPayload
	if err := decodeJSON(r, &payload); err != nil {
		response.Error(w, r, err)
		return
	}

	result, err := h.deps.Subscriptions.ApplyImport(r.Context(), payload)
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.Data(w, http.StatusOK, result)
}

File: frontend/src/lib/api/subscriptions.ts
/** Artist subscriptions: watching an artist and syncing their discography. */

import { request, requestList, requestVoid } from '@/lib/api/client'
import type {
  ImportPreview,
  ImportResult,
  SubscribeRequest,
  Subscription,
  SubscriptionDetail,
  SubscriptionExport,
  SubscriptionUpdate,
  SyncStatus,
} from '@/types/api'

/** GET /subscriptions */
export async function listSubscriptions(
  options: {
    provider?: string
    artistSourceId?: string
    limit?: number
    signal?: AbortSignal
  } = {},
): Promise<Subscription[]> {
  const result = await requestList<Subscription>('/subscriptions', {
    query: {
      provider: options.provider,
      artist_source_id: options.artistSourceId,
      limit: options.limit,
    },
    signal: options.signal,
  })
  return result.items
}

/**
 * The subscription of one artist, or null when the artist is not watched.
 *
 * This is what the artist page asks. It goes through the list filter rather
 * than a lookup endpoint of its own, because "not subscribed" is a normal
 * answer and must not arrive as a 404 the caller has to catch.
 */
export async function findSubscription(
  provider: string,
  artistSourceId: string,
  signal?: AbortSignal,
): Promise<Subscription | null> {
  const list = await listSubscriptions({ provider, artistSourceId, signal })
  return list[0] ?? null
}

/**
 * POST /subscriptions
 *
 * Subscribing to an artist that is already watched returns the existing
 * subscription, so a double click is harmless.
 */
export function subscribe(
  body: SubscribeRequest,
  signal?: AbortSignal,
): Promise<Subscription> {
  return request<Subscription>('/subscriptions', { method: 'POST', body, signal })
}

/** GET /subscriptions/{id} — the subscription plus the last report, if any. */
export function getSubscription(
  id: string,
  signal?: AbortSignal,
): Promise<SubscriptionDetail> {
  return request<SubscriptionDetail>(`/subscriptions/${encodeURIComponent(id)}`, {
    signal,
  })
}

/** PATCH /subscriptions/{id} — an omitted field is left untouched. */
export function updateSubscription(
  id: string,
  body: SubscriptionUpdate,
  signal?: AbortSignal,
): Promise<Subscription> {
  return request<Subscription>(`/subscriptions/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    body,
    signal,
  })
}

/**
 * DELETE /subscriptions/{id}
 *
 * Only the subscription goes. Everything that was already downloaded stays in
 * the library — a subscription says what to watch, not what to keep.
 */
export function unsubscribe(id: string, signal?: AbortSignal): Promise<void> {
  // The backend answers 204 with no body, so this must not go through
  // request(), which requires a "data" envelope.
  return requestVoid(`/subscriptions/${encodeURIComponent(id)}`, {
    method: 'DELETE',
    signal,
  })
}

/**
 * POST /subscriptions/{id}/sync
 *
 * The backend answers 202 and walks the discography in the background, so the
 * answer means "started", never "finished". Progress arrives on the event
 * stream; the report is read back with getSubscription.
 */
export function syncSubscription(
  id: string,
  signal?: AbortSignal,
): Promise<Subscription> {
  return request<Subscription>(`/subscriptions/${encodeURIComponent(id)}/sync`, {
    method: 'POST',
    signal,
  })
}

/** GET /subscriptions/export — download all subscriptions in portable JSON format. */
export function getSubscriptionExport(signal?: AbortSignal): Promise<SubscriptionExport> {
  return request<SubscriptionExport>('/subscriptions/export', { signal })
}

/** Triggers a browser download of any JSON payload. */
export function downloadJsonFile(data: unknown, filename: string): void {
  const jsonStr = JSON.stringify(data, null, 2)
  const blob = new Blob([jsonStr], { type: 'application/json' })
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  document.body.appendChild(a)
  a.click()
  document.body.removeChild(a)
  URL.revokeObjectURL(url)
}

/** Export all subscriptions or a selected subset to a downloadable JSON file. */
export async function exportSubscriptionsToFile(selectedSubs?: Subscription[]): Promise<void> {
  const dateStr = new Date().toISOString().slice(0, 10)
  if (selectedSubs && selectedSubs.length > 0) {
    const payload: SubscriptionExport = {
      format: 'ytmdl-subscriptions',
      version: 1,
      exported_at: new Date().toISOString(),
      subscriptions: selectedSubs.map((s) => ({
        artist_name: s.artist_name,
        provider: s.provider,
        artist_source_id: s.artist_source_id,
        artist_image_url: s.artist_image_url,
        enabled: s.enabled,
        auto_download: s.auto_download,
        release_filter: s.release_filter,
        download_priority: s.download_priority,
      })),
    }
    downloadJsonFile(payload, `ytmdl-subscriptions-selected-${dateStr}.json`)
    return
  }
  const data = await getSubscriptionExport()
  downloadJsonFile(data, `ytmdl-subscriptions-${dateStr}.json`)
}

/** POST /subscriptions/import/preview — analyze subscriptions file without DB changes. */
export function previewImportSubscriptions(
  body: SubscriptionExport,
  signal?: AbortSignal,
): Promise<ImportPreview> {
  return request<ImportPreview>('/subscriptions/import/preview', {
    method: 'POST',
    body,
    signal,
  })
}

/** POST /subscriptions/import/apply — apply imported subscriptions in a transaction. */
export function applyImportSubscriptions(
  body: SubscriptionExport,
  signal?: AbortSignal,
): Promise<ImportResult> {
  return request<ImportResult>('/subscriptions/import/apply', {
    method: 'POST',
    body,
    signal,
  })
}

/* ------------------------------------------------------------------ helpers */

/** The German label of a sync status. */
export const SYNC_STATUS_LABELS: Record<SyncStatus, string> = {
  pending: 'Noch nicht geprüft',
  success: 'Erfolgreich',
  partial: 'Teilweise',
  failed: 'Fehlgeschlagen',
}

/** The badge tone a sync status is shown with. */
export function syncStatusTone(
  status: SyncStatus,
): 'neutral' | 'success' | 'warning' | 'destructive' {
  switch (status) {
    case 'success':
      return 'success'
    case 'partial':
      return 'warning'
    case 'failed':
      return 'destructive'
    default:
      return 'neutral'
  }
}

/**
 * The one-line summary of a finished run.
 *
 * A run that found nothing says so: "0 neue Releases · 0 neue Tracks" reads as
 * a failure at a glance, while "Keine Neuigkeiten" is what actually happened.
 */
export function summarizeSync(result: {
  new_releases: number
  new_tracks: number
  queued_tracks: number
}): string {
  const parts: string[] = []
  if (result.new_releases > 0) {
    parts.push(
      result.new_releases === 1
        ? '1 neues Release'
        : `${result.new_releases} neue Releases`,
    )
  }
  if (result.new_tracks > 0) {
    parts.push(
      result.new_tracks === 1 ? '1 neuer Track' : `${result.new_tracks} neue Tracks`,
    )
  }
  if (parts.length === 0) return 'Keine Neuigkeiten'

  if (result.queued_tracks > 0) {
    parts.push(
      result.queued_tracks === 1
        ? '1 Track in der Warteschlange'
        : `${result.queued_tracks} Tracks in der Warteschlange`,
    )
  }
  return parts.join(' · ')
}

/** Whether a subscription is due for its next scheduled run. */
export function isDue(subscription: Subscription, now: Date = new Date()): boolean {
  if (!subscription.enabled) return false
  const next = new Date(subscription.next_sync_at)
  if (Number.isNaN(next.getTime())) return false
  return next.getTime() <= now.getTime()
}
